src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            TypeSyntax declaredType;
+            State state = null;
+            var shouldAnalyze = false;
+            var declarationStatement = context.Node;
+            var optionSet = GetOptionSet(context.Options);
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+
+            if (declarationStatement.IsKind(SyntaxKind.VariableDeclaration))
+            {
+                var declaration = (VariableDeclarationSyntax)declarationStatement;
+                declaredType = declaration.Type;
+
+                shouldAnalyze = ShouldAnalyzeVariableDeclaration(declaration, semanticModel, cancellationToken);
+
+                if (shouldAnalyze)
+                {
+                    state = State.Generate(declarationStatement, semanticModel, optionSet, isVariableDeclarationContext: true, cancellationToken: cancellationToken);
+                    shouldAnalyze = IsStylePreferred(semanticModel, optionSet, state, cancellationToken);
+                }
+            }
+            else if (declarationStatement.IsKind(SyntaxKind.ForEachStatement))
+            {
+                var declaration = (ForEachStatementSyntax)declarationStatement;
+                declaredType = declaration.Type;
+
+                state = State.Generate(declarationStatement, semanticModel, optionSet, isVariableDeclarationContext: false, cancellationToken: cancellationToken);
+                shouldAnalyze = IsStylePreferred(semanticModel, optionSet, state, cancellationToken);
+            }
+            else
+            {
+                Debug.Assert(false, $"called in for unregistered node kind {declarationStatement.Kind().ToString()}");
+                return;
+            }
+
+            if (shouldAnalyze)
+            {
+                Debug.Assert(state != null, "analyzing a declaration and state is null.");
+
+                TextSpan diagnosticSpan;
+
+                if (TryAnalyzeVariableDeclaration(declaredType, semanticModel, optionSet, cancellationToken, out diagnosticSpan))
+                {
+                    var descriptor = _severityToDescriptorMap[state.GetDiagnosticSeverityPreference()];
+                    context.ReportDiagnostic(CreateDiagnostic(descriptor, declarationStatement, diagnosticSpan));
+                }
+            }
+        }
+
+        private Diagnostic CreateDiagnostic(DiagnosticDescriptor descriptor, SyntaxNode declaration, TextSpan diagnosticSpan) =>
+            Diagnostic.Create(descriptor, declaration.SyntaxTree.GetLocation(diagnosticSpan));
+
+        private bool ShouldAnalyzeVariableDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            // var is applicable only for local variables.
+            if (variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration,
```

Comment:
the problem with a blacklist is that you might forget something (and then who knows what happens).  With a whitelist you specify precisely what you handle.  Please whitelist here.